Mathlib/FieldTheory/IsPerfectClosure.lean

```diff
@@ -0,0 +1,497 @@
+      liftAux_apply i j p (x1 * x2) (n1 + n2) (y1 ^ p ^ n2 * y2 ^ p ^ n1) (by rw [map_mul,
+        map_pow, map_pow, h1, h2, ← pow_mul, ← pow_add, ← pow_mul, ← pow_add,
+        add_comm n2, mul_pow]),
+      map_mul, map_pow, map_pow, map_mul, ← iterateFrobeniusEquiv_def]
+    nth_rw 1 [iterateFrobeniusEquiv_symm_add_apply]
+    rw [RingEquiv.symm_apply_apply, add_comm n1, iterateFrobeniusEquiv_symm_add_apply,
+      ← iterateFrobeniusEquiv_def, RingEquiv.symm_apply_apply]
+  map_zero' := by simp [liftAux_apply i j p 0 0 0 (by rw [pow_zero, pow_one, map_zero])]
+  map_add' x1 x2 := by
+    obtain ⟨n1, y1, h1⟩ := IsPRadical.pow_mem i p x1
+    obtain ⟨n2, y2, h2⟩ := IsPRadical.pow_mem i p x2
+    simp only; rw [liftAux_apply i j p _ _ _ h1, liftAux_apply i j p _ _ _ h2,
+      liftAux_apply i j p (x1 + x2) (n1 + n2) (y1 ^ p ^ n2 + y2 ^ p ^ n1) (by rw [map_add,
+        map_pow, map_pow, h1, h2, ← pow_mul, ← pow_add, ← pow_mul, ← pow_add,
+        add_comm n2, add_pow_expChar_pow]),
+      map_add, map_pow, map_pow, map_add, ← iterateFrobeniusEquiv_def]
+    nth_rw 1 [iterateFrobeniusEquiv_symm_add_apply]
+    rw [RingEquiv.symm_apply_apply, add_comm n1, iterateFrobeniusEquiv_symm_add_apply,
+      ← iterateFrobeniusEquiv_def, RingEquiv.symm_apply_apply]
+
+theorem PerfectRing.lift_apply (x : L) (n : ℕ) (y : K) (h : i y = x ^ p ^ n) :
+    lift i j p x = (iterateFrobeniusEquiv M p n).symm (j y) :=
+  liftAux_apply i j p _ _ _ h
+
+theorem PerfectRing.lift_self_apply [PerfectRing L p] (x : L) : lift i i p x = x :=
+  liftAux_self_apply i p x
+
+theorem PerfectRing.lift_self [PerfectRing L p] : lift i i p = RingHom.id L :=
+  RingHom.ext (liftAux_self_apply i p)
+
+theorem PerfectRing.lift_id_apply (x : K) : lift (RingHom.id K) j p x = j x :=
+  liftAux_id_apply j p x
+
+theorem PerfectRing.lift_id : lift (RingHom.id K) j p = j :=
+  RingHom.ext (liftAux_id_apply j p)
+
+theorem PerfectRing.lift_comp_apply (x : K) : lift i j p (i x) = j x := by
+  rw [lift_apply i j p _ 0 x (by rw [pow_zero, pow_one]), iterateFrobeniusEquiv_zero]; rfl
+
+theorem PerfectRing.lift_comp : (lift i j p).comp i = j :=
+  RingHom.ext (lift_comp_apply i j p)
+
+theorem PerfectRing.comp_lift_apply (x : L) : lift i (f.comp i) p x = f x := by
+  obtain ⟨n, y, h⟩ := IsPRadical.pow_mem i p x
+  rw [lift_apply i _ p _ _ _ h, RingHom.comp_apply, h, ← iterate_frobenius, f.map_iterate_frobenius,
+    ← coe_iterateFrobenius, ← iterateFrobeniusEquiv_apply, RingEquiv.symm_apply_apply]
+
+theorem PerfectRing.comp_lift : lift i (f.comp i) p = f :=
+  RingHom.ext (comp_lift_apply i f p)
+
+variable (M) in
+/-- If `i : K →+* L` is a homomorphisms of characteristic `p` rings, such that
+`i` is `p`-radical, and `M` is a perfect ring of characteristic `p`,
+then `K →+* M` is one-to-one correspondence to
+`L →+* M`, given by `PerfectRing.lift`. This is a generalization to `PerfectClosure.lift`. -/
+def PerfectRing.liftEquiv : (K →+* M) ≃ (L →+* M) where
```

Comment:
After this definition, you should add the following lemma (not sure it is the best name though):
```lean
lemma RingHom.ext_of_perfectRing
    {φ₁ φ₂ : L →+* M} (h : φ₁.comp i = φ₂.comp i) : φ₁ = φ₂ :=
  (PerfectRing.liftEquiv M i p).symm.injective h
```
Then, using the simp lemma `lift_comp_apply`, some proofs should be facilitated, like:
```lean
theorem PerfectRing.lift_comp_lift [PerfectRing N p] [IsPRadical j p] :
    (lift j k p).comp (lift i j p) = lift i k p :=
  RingHom.ext_of_perfectRing i p (by ext; simp)
```

Reply:
I think your `RingHom.ext_of_perfectRing` is closely related to [IsPurelyInseparable.injective_comp_algebraMap](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/PurelyInseparable.html#IsPurelyInseparable.injective_comp_algebraMap). Let me study if a `IsPRadical.injective_comp` should be added.

Reply:
Thanks for the suggestion. But I only spot `PerfectRing.lift_comp_lift` which can be benefitted by this result. Are there any other?

Reply:
Some lemmas about `liftAux` you mentioned only requires `CommSemiring`, however `IsPRadical.injective_comp_of_perfect` requires `CommRing`. So personally I don't want to remove them. Ohterwise I think your suggestion is good.

Reply:
I spotted another lemma `lift_lift` which can be golfed.